File: packages/skill-sdk-typescript/src/context.ts
export type PrincipalType = "user" | "admin" | "service" | "agent" | "event" | "schedule";

export type SkillPrincipal = Readonly<{
  principalType: PrincipalType;
  principalId: string;
  organizationId?: string;
}>;

export type SkillContext = Readonly<{
  executionId: string;
  installationId: string;
  actorUserId?: string;
  principal: SkillPrincipal;
  locale: string;
  timezone: string;
  idempotencyKey?: string;
  deadline: Date;
  permissions: ReadonlySet<string>;
  capabilityGrants: ReadonlySet<string>;
  requestId: string;
  traceId: string;
  signal: AbortSignal;
}>;

export function assertContextActive(context: SkillContext, now = new Date()): void {
  if (context.signal.aborted) throw new SkillCancelledError();
  if (now.getTime() >= context.deadline.getTime()) throw new SkillDeadlineExceededError();
}

export class SkillCancelledError extends Error {
  readonly code = "SKILL_CANCELLED";
  constructor() { super("SKILL_CANCELLED"); }
}

export class SkillDeadlineExceededError extends Error {
  readonly code = "SKILL_DEADLINE_EXCEEDED";
  constructor() { super("SKILL_DEADLINE_EXCEEDED"); }
}

File: packages/skill-sdk-python/src/vav_skill_sdk/context.py
"""Immutable execution identity and authority passed to Skills."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator


class SkillPrincipal(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    principal_type: Literal["user", "admin", "service", "agent", "event", "schedule"]
    principal_id: str = Field(min_length=1, max_length=255)
    organization_id: UUID | None = None


class SkillContext(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    execution_id: UUID
    installation_id: UUID
    actor_user_id: UUID | None = None
    principal: SkillPrincipal
    locale: str = Field(pattern=r"^[a-z]{2}(?:-[A-Z]{2})?$")
    timezone: str = Field(min_length=1, max_length=64)
    idempotency_key: str | None = Field(default=None, min_length=8, max_length=128)
    deadline: datetime
    permissions: frozenset[str] = frozenset()
    capability_grants: frozenset[str] = frozenset()
    request_id: UUID
    trace_id: str = Field(pattern=r"^[0-9a-f]{16,64}$")

    @model_validator(mode="after")
    def deadline_has_timezone(self) -> "SkillContext":
        if self.deadline.tzinfo is None:
            raise ValueError("deadline must be timezone-aware")
        return self
